File: contracts/rust/inccounter/inccounter.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package inccounter

import "github.com/iotaledger/wasp/packages/vm/wasmlib"

var LocalStateMustIncrement = false

func funcCallIncrement(ctx *wasmlib.ScFuncContext, params *FuncCallIncrementParams) {
	counter := ctx.State().GetInt(VarCounter)
	value := counter.Value()
	counter.SetValue(value + 1)
	if value == 0 {
		ctx.CallSelf(HFuncCallIncrement, nil, nil)
	}
}

func funcCallIncrementRecurse5x(ctx *wasmlib.ScFuncContext, params *FuncCallIncrementRecurse5xParams) {
	counter := ctx.State().GetInt(VarCounter)
	value := counter.Value()
	counter.SetValue(value + 1)
	if value < 5 {
		ctx.CallSelf(HFuncCallIncrementRecurse5x, nil, nil)
	}
}

func funcIncrement(ctx *wasmlib.ScFuncContext, params *FuncIncrementParams) {
	counter := ctx.State().GetInt(VarCounter)
	counter.SetValue(counter.Value() + 1)
}

func funcInit(ctx *wasmlib.ScFuncContext, params *FuncInitParams) {
	counter := params.Counter.Value()
	if counter == 0 {
		return
	}
	ctx.State().GetInt(VarCounter).SetValue(counter)
}

func funcLocalStateInternalCall(ctx *wasmlib.ScFuncContext, params *FuncLocalStateInternalCallParams) {
	LocalStateMustIncrement = false
	par := &FuncWhenMustIncrementParams{}
	funcWhenMustIncrement(ctx, par)
	LocalStateMustIncrement = true
	funcWhenMustIncrement(ctx, par)
	funcWhenMustIncrement(ctx, par)
	// counter ends up as 2
}

func funcLocalStatePost(ctx *wasmlib.ScFuncContext, params *FuncLocalStatePostParams) {
	LocalStateMustIncrement = false
	request := &wasmlib.PostRequestParams{
		ContractId: ctx.ContractId(),
		Function:   HFuncWhenMustIncrement,
		Params:     nil,
		Transfer:   nil,
		Delay:      0,
	}
	ctx.Post(request)
	LocalStateMustIncrement = true
	ctx.Post(request)
	ctx.Post(request)
	// counter ends up as 0 (non-existent)
}

func funcLocalStateSandboxCall(ctx *wasmlib.ScFuncContext, params *FuncLocalStateSandboxCallParams) {
	LocalStateMustIncrement = false
	ctx.CallSelf(HFuncWhenMustIncrement, nil, nil)
	LocalStateMustIncrement = true
	ctx.CallSelf(HFuncWhenMustIncrement, nil, nil)
	ctx.CallSelf(HFuncWhenMustIncrement, nil, nil)
	// counter ends up as 0 (non-existent)
}

func funcPostIncrement(ctx *wasmlib.ScFuncContext, params *FuncPostIncrementParams) {
	counter := ctx.State().GetInt(VarCounter)
	value := counter.Value()
	counter.SetValue(value + 1)
	if value == 0 {
		ctx.Post(&wasmlib.PostRequestParams{
			ContractId: ctx.ContractId(),
			Function:   HFuncPostIncrement,
			Params:     nil,
			Transfer:   nil,
			Delay:      0,
		})
	}
}

func funcRepeatMany(ctx *wasmlib.ScFuncContext, params *FuncRepeatManyParams) {
	counter := ctx.State().GetInt(VarCounter)
	value := counter.Value()
	counter.SetValue(value + 1)
	stateRepeats := ctx.State().GetInt(VarNumRepeats)
	repeats := params.NumRepeats.Value()
	if repeats == 0 {
		repeats = stateRepeats.Value()
		if repeats == 0 {
			return
		}
	}
	stateRepeats.SetValue(repeats - 1)
	ctx.Post(&wasmlib.PostRequestParams{
		ContractId: ctx.ContractId(),
		Function:   HFuncRepeatMany,
		Params:     nil,
		Transfer:   nil,
		Delay:      0,
	})
}

func funcResultsTest(ctx *wasmlib.ScFuncContext, params *FuncResultsTestParams) {
	testMap(ctx.Results())
	checkMap(ctx.Results().Immutable())
	//ctx.CallSelf(HFuncResultsCheck, nil, nil)
}

func funcStateTest(ctx *wasmlib.ScFuncContext, params *FuncStateTestParams) {
	testMap(ctx.State())
	ctx.CallSelf(HViewStateCheck, nil, nil)
}

func funcWhenMustIncrement(ctx *wasmlib.ScFuncContext, params *FuncWhenMustIncrementParams) {
	ctx.Log("when_must_increment called")
	{
		if !LocalStateMustIncrement {
			return
		}
	}
	counter := ctx.State().GetInt(VarCounter)
	counter.SetValue(counter.Value() + 1)
}

func viewGetCounter(ctx *wasmlib.ScViewContext, params *ViewGetCounterParams) {
	counter := ctx.State().GetInt(VarCounter).Value()
	ctx.Results().GetInt(VarCounter).SetValue(counter)
}

func viewResultsCheck(ctx *wasmlib.ScViewContext, params *ViewResultsCheckParams) {
	checkMap(ctx.Results().Immutable())
}

func viewStateCheck(ctx *wasmlib.ScViewContext, params *ViewStateCheckParams) {
	checkMap(ctx.State())
}

func testMap(kvstore wasmlib.ScMutableMap) {
	int1 := kvstore.GetInt(VarInt1)
	check(int1.Value() == 0)
	int1.SetValue(1)

	string1 := kvstore.GetString(VarString1)
	check(string1.Value() == "")
	string1.SetValue("a")
	check(string1.Value() == "a")

	ia1 := kvstore.GetIntArray(VarIntArray1)
	int2 := ia1.GetInt(0)
	check(int2.Value() == 0)
	int2.SetValue(2)
	int3 := ia1.GetInt(1)
	check(int3.Value() == 0)
	int3.SetValue(3)

	sa1 := kvstore.GetStringArray(VarStringArray1)
	string2 := sa1.GetString(0)
	check(string2.Value() == "")
	string2.SetValue("bc")
	string3 := sa1.GetString(1)
	check(string3.Value() == "")
	string3.SetValue("def")
}

func checkMap(kvstore wasmlib.ScImmutableMap) {
	int1 := kvstore.GetInt(VarInt1)
	check(int1.Value() == 1)

	string1 := kvstore.GetString(VarString1)
	check(string1.Value() == "a")

	ia1 := kvstore.GetIntArray(VarIntArray1)
	int2 := ia1.GetInt(0)
	check(int2.Value() == 2)
	int3 := ia1.GetInt(1)
	check(int3.Value() == 3)

	sa1 := kvstore.GetStringArray(VarStringArray1)
	string2 := sa1.GetString(0)
	check(string2.Value() == "bc")
	string3 := sa1.GetString(1)
	check(string3.Value() == "def")
}

func check(condition bool) {
	if !condition {
		panic("Check failed!")
	}
}

File: contracts/rust/inccounter/src/inccounter.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use wasmlib::*;

use crate::*;

static mut LOCAL_STATE_MUST_INCREMENT: bool = false;

pub fn func_call_increment(ctx: &ScFuncContext) {
    let counter = ctx.state().get_int64(VAR_COUNTER);
    let value = counter.value();
    counter.set_value(value + 1);
    if value == 0 {
        ctx.call_self(HFUNC_CALL_INCREMENT, None, None);
    }
}

pub fn func_call_increment_recurse5x(ctx: &ScFuncContext) {
    let counter = ctx.state().get_int64(VAR_COUNTER);
    let value = counter.value();
    counter.set_value(value + 1);
    if value < 5 {
        ctx.call_self(HFUNC_CALL_INCREMENT_RECURSE5X, None, None);
    }
}

pub fn func_increment(ctx: &ScFuncContext) {
    let counter = ctx.state().get_int64(VAR_COUNTER);
    counter.set_value(counter.value() + 1);
}

pub fn func_init(ctx: &ScFuncContext) {
    let p = ctx.params();
    let param_counter = p.get_int64(PARAM_COUNTER);
    if param_counter.exists() {
        let counter = param_counter.value();
        ctx.state().get_int64(VAR_COUNTER).set_value(counter);
    }
}

pub fn func_local_state_internal_call(ctx: &ScFuncContext) {
    unsafe {
        LOCAL_STATE_MUST_INCREMENT = false;
    }
    func_when_must_increment(ctx);
    unsafe {
        LOCAL_STATE_MUST_INCREMENT = true;
    }
    func_when_must_increment(ctx);
    func_when_must_increment(ctx);
    // counter ends up as 2
}

pub fn func_local_state_post(ctx: &ScFuncContext) {
    unsafe {
        LOCAL_STATE_MUST_INCREMENT = false;
    }
    // prevent multiple identical posts, need a dummy param to differentiate them
    local_state_post(ctx, 1);
    unsafe {
        LOCAL_STATE_MUST_INCREMENT = true;
    }
    local_state_post(ctx, 2);
    local_state_post(ctx, 3);
    // counter ends up as 0
}

pub fn func_loop(_ctx: &ScFuncContext) {
    loop {}
}

fn local_state_post(ctx: &ScFuncContext, nr: i64) {
    let params = ScMutableMap::new();
    params.get_int64(VAR_INT1).set_value(nr);
    let transfer = ScTransfers::iotas(1);
    ctx.post_self(HFUNC_WHEN_MUST_INCREMENT, Some(params), transfer, 0);
}

pub fn func_local_state_sandbox_call(ctx: &ScFuncContext) {
    unsafe {
        LOCAL_STATE_MUST_INCREMENT = false;
    }
    ctx.call_self(HFUNC_WHEN_MUST_INCREMENT, None, None);
    unsafe {
        LOCAL_STATE_MUST_INCREMENT = true;
    }
    ctx.call_self(HFUNC_WHEN_MUST_INCREMENT, None, None);
    ctx.call_self(HFUNC_WHEN_MUST_INCREMENT, None, None);
    // counter ends up as 0
}

pub fn func_post_increment(ctx: &ScFuncContext) {
    let counter = ctx.state().get_int64(VAR_COUNTER);
    let value = counter.value();
    counter.set_value(value + 1);
    if value == 0 {
        let transfer = ScTransfers::iotas(1);
        ctx.post_self(HFUNC_POST_INCREMENT, None, transfer, 0);
    }
}

pub fn func_repeat_many(ctx: &ScFuncContext) {
    let p = ctx.params();
    let param_num_repeats = p.get_int64(PARAM_NUM_REPEATS);

    let counter = ctx.state().get_int64(VAR_COUNTER);
    let value = counter.value();
    counter.set_value(value + 1);
    let state_repeats = ctx.state().get_int64(VAR_NUM_REPEATS);
    let mut repeats = param_num_repeats.value();
    if repeats == 0 {
        repeats = state_repeats.value();
        if repeats == 0 {
            return;
        }
    }
    state_repeats.set_value(repeats - 1);
    let transfer = ScTransfers::iotas(1);
    ctx.post_self(HFUNC_REPEAT_MANY, None, transfer, 0);
}

pub fn func_when_must_increment(ctx: &ScFuncContext) {
    ctx.log("when_must_increment called");
    unsafe {
        if !LOCAL_STATE_MUST_INCREMENT {
            return;
        }
    }
    let counter = ctx.state().get_int64(VAR_COUNTER);
    counter.set_value(counter.value() + 1);
}

// note that get_counter mirrors the state of the 'counter' state variable
// which means that if the state variable was not present it also will not be present in the result
pub fn view_get_counter(ctx: &ScViewContext) {
    let counter = ctx.state().get_int64(VAR_COUNTER);
    if counter.exists() {
        ctx.results().get_int64(VAR_COUNTER).set_value(counter.value());
    }
}

pub fn func_test_leb128(ctx: &ScFuncContext) {
    save(ctx, "v-1", -1);
    save(ctx, "v-2", -2);
    save(ctx, "v-126", -126);
    save(ctx, "v-127", -127);
    save(ctx, "v-128", -128);
    save(ctx, "v-129", -129);
    save(ctx, "v0", 0);
    save(ctx, "v+1", 1);
    save(ctx, "v+2", 2);
    save(ctx, "v+126", 126);
    save(ctx, "v+127", 127);
    save(ctx, "v+128", 128);
    save(ctx, "v+129", 129);
}

fn save(ctx: &ScFuncContext, name: &str, value: i64) {
    let mut encoder = BytesEncoder::new();
    encoder.int64(value);
    let spot = ctx.state().get_bytes(name);
    spot.set_value(&encoder.data());

    let bytes = spot.value();
    let mut decoder = BytesDecoder::new(&bytes);
    let retrieved = decoder.int64();
    if retrieved != value {
        ctx.log(&(name.to_string() + " in : " + &value.to_string()));
        ctx.log(&(name.to_string() + " out: " + &retrieved.to_string()));
    }
}
